### PR TITLE
Events tracking + Segment

### PR DESCRIPTION
**Pull request by JoshBroomberg**
TO DO
- Finalise a list of user attributes to track
- Decide on when to do updates to user. (Is after save correct?)
- Decide on which events to track and build them out (very quick)

This is the foundational PR for the events service. It introduces:

**Objects**
- An `Event` model which is:
 - Owned by an eventable
 - Has an event type enum
 - Has a json attribute called data

- Event submodels for each event type.

 These event models define custom event data via non_persisted active type attributes. A clever little before_save in the base event class collects all the values from these AT attributes and stores them into the json data attribute before save.  *This allows much cleaner event object instantiation, allowing things like validations on the custom data. The alternative is to pass in a json hash to the event from the event service. Much messier* 

 (Check out the `written_application_started` event for custom data - none is present  in `sign_up`)

- An `EventService` to wrap the creation of events.

**Flow**
- Controller calls `EventService` method to register an event, passing in required models. 
- Event service creates a new event of the relevant class, filling in both common persisted attributes and the custom ActiveType attributes. 
- The base event class uses a before_save to collect together all the non_persisted custom data and save it to the json `data` field.
- The base event class stores the event instance to be tracked via the front end.

**Summary**
 This system is in place and working like a charm. Look forward to your comments.

### CODE REVIEW

File: app/models/journey.rb [hunk 1]
```diff
@@ -17,4 +18,11 @@ def sanitize_journey_stages
       user.applications.started.destroy_all
     end
   end
+
+  def track_applications_started
+    journey_stages.each do |stage|
+      EventService.written_application_started(
+        user, stage.session.program)
```
Comment (JoshBroomberg) [inline app/models/journey.rb]: Not super happy with how many queries this involves.

File: app/models/event.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
+class Event < ActiveRecord::Base
+  belongs_to :eventable, polymorphic: true
```
Comment (khanotations) [inline app/models/event.rb]: I thought we agreed all events belonged to users?

File: app/models/event/written_application_started.rb [hunk 3]
```diff
@@ -0,0 +1,6 @@
+class Event::WrittenApplicationStarted < ActiveType::Record[Event]
+   attribute :program_name, :string
```
Comment (khanotations) [inline app/models/event/written_application_started.rb]: This is not sufficient; how do we handle multiple programs?

File: app/models/journey.rb [hunk 4]
```diff
@@ -17,4 +18,11 @@ def sanitize_journey_stages
       user.applications.started.destroy_all
     end
   end
+
+  def track_applications_started
```
Comment (khanotations) [inline app/models/journey.rb]: Is `track` going to be the word we use for all event generating callbacks? We should pick a verb and stick to ti

File: app/models/journey.rb (hunk 1 above)
Comment (khanotations) [inline app/models/journey.rb]: Yeah shouldn't we generate one event that has multiple courses? And what if someone changes their course?

File: app/models/event/sign_up.rb [hunk 5]
```diff
@@ -0,0 +1,3 @@
+class Event::SignUp < ActiveType::Record[Event]
+   validates_inclusion_of :event_type, in: ["sign_up"]
```
Comment (khanotations) [inline app/models/event/sign_up.rb]: I think like with `Question::AsWritten`, we should have `default_scope { sign_up }`

File: app/services/event_service.rb [hunk 6]
```diff
@@ -0,0 +1,20 @@
+class EventService
+  def self.sign_up(user)
+    Event::SignUp.create!(
+      eventable: user,
+      event_type: event_type_value(:sign_up))
```
Comment (khanotations) [inline app/services/event_service.rb]: If we add the `default_scope`, this `event_type` becomes unnecessary.

File: app/services/event_service.rb [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+  end
+
+  def self.written_application_started(user, program)
+    Event::WrittenApplicationStarted.create!(
+      eventable: user,
+      event_type: event_type_value(:written_application_started),
```
Comment (khanotations) [inline app/services/event_service.rb]: same here

File: app/services/event_service.rb [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+      program_name: program.name)
+  end
+
+  private
+
+  def self.event_type_value(event_name)
```
Comment (khanotations) [inline app/services/event_service.rb]: lol this is also unnecessary.

File: app/models/event.rb [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+  before_save :assign_json_from_attributes
+  after_create :sync_event
+
+  private
+
+  # Programatically determines all non-persisted columns
+  # (which serve to store event attribute data)
+  # and converts them to json to be stored in 'data' 
+  # before saving.
+  def assign_json_from_attributes
```
Comment (khanotations) [inline app/models/event.rb]: This seems a little weird. It seems like we should just do `event.attributes.x = y`, and it would store automatically, no?

File: app/services/event_service.rb [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+    Event::SignUp.create!(
+      eventable: user,
+      event_type: event_type_value(:sign_up))
+  end
+
+  def self.written_application_started(user, program)
+    Event::WrittenApplicationStarted.create!(
+      eventable: user,
+      event_type: event_type_value(:written_application_started),
+      program_name: program.name)
```
Comment (khanotations) [inline app/services/event_service.rb]: so here, write `attributes.program_name: program.name`? Does that work?

File: app/models/journey.rb (hunk 4 above)
Comment (JoshBroomberg) [inline app/models/journey.rb]: I like track. Let's make it the convention.

File: app/services/event_service.rb (hunk 6 above)
Comment (JoshBroomberg) [inline app/services/event_service.rb]: does default scope affect new models?

File: app/models/journey.rb (hunk 1 above)
Comment (JoshBroomberg) [inline app/models/journey.rb]: No, if you do that it becomes very annoying to use the event because you have to query the program_name by contains instead of by equals due to many possible combos. Separate events, with different data is the way to go I think. 

Intercom encourages this by giving you the ability to act on the first occurrence of a given event, meaning you don't have to smush all your data into one event.

As to change course, it will just register a different course as having been started which is logically accurate. Started Web Dev with internship, then later started Data science and then submitted data science.

Comment (JoshBroomberg) [inline app/models/journey.rb]: Hmm, maybe we could concat program names?

Comment (JoshBroomberg) [inline app/models/journey.rb]: IDK

File: app/models/event.rb (hunk 2 above)
Comment (JoshBroomberg) [inline app/models/event.rb]: I saw no real reason not to include this flexibility. If we only ever link them to users, we suffer nothing, if we do ever want to track events on other models, we thank our lucky stars

Comment (JoshBroomberg) [inline app/models/event.rb]: Unless we just convert to eventable at that time

File: app/models/event/written_application_started.rb (hunk 3 above)
Comment (khanotations) [inline app/models/event/written_application_started.rb]: Add `default_scope`

**Comment by JoshBroomberg** [comment]
@khanotations all changed made. Looking really nice. Check out the segment code. Super clean.

File: app/models/event.rb [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+    :written_application_submitted,
+    :interview_submitted
+  ]
+  
+  before_save :assign_json_from_attributes
+  after_create :sync_event
+
+  private
+
+  # Programatically determines all non-persisted columns
```
Comment (khanotations) [inline app/models/event.rb]: Please include in this note that we jump through this hoop in order to allow the use of AT attributes

File: app/models/interview_submission.rb [hunk 12]
```diff
@@ -23,6 +24,10 @@ def create_interview_response_questions
     end
   end
 
+  def track_interview_submitted
+    EventService.interview_submitted(interview_response.submitted_user)
```
Comment (khanotations) [inline app/models/interview_submission.rb]: can we delegate `submitted_user` to `interview_response` in this class? I think it makes sense that `interview_submission` has a method called `submitted_user`

File: app/models/journey.rb [hunk 13]
```diff
@@ -18,6 +19,11 @@ def sanitize_journey_stages
     end
   end
 
+  def track_applications_started
+    sessions = journey_stages.map(&:session)
```
Comment (khanotations) [inline app/models/journey.rb]: similarly, should `sessions` be a method in this class? `journey.sessions` makes sense to me.

File: app/models/user/as_segmentable.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+class User::AsSegmentable < ActiveType::Record[User]
+  
+  def sync
+    Analytics.identify(user_id: "#{id}", traits: traits_hash)
```
Comment (khanotations) [inline app/models/user/as_segmentable.rb]: why is this string interpolated..?

File: app/models/user/as_segmentable.rb [hunk 15]
```diff
@@ -0,0 +1,16 @@
...
+  end
+
+  private
+
+  def traits_hash
+    {
```
Comment (khanotations) [inline app/models/user/as_segmentable.rb]: Any idea what happens when email/first name etc changes?

File: app/models/user.rb [hunk 16]
```diff
@@ -33,4 +36,8 @@ def email_required?
   def password_required?
     false
   end
+
+  def sync_user
+    self.as(User::AsSegmentable).sync
```
Comment (khanotations) [inline app/models/user.rb]: Are we sure we want to commit to this...? This is after every save. I don't see this in the docs.

File: app/models/user/as_segmentable.rb (hunk 14 above)
Comment (khanotations) [inline app/models/user/as_segmentable.rb]: Also in the docs it says to create an instance of `Analytics`

File: app/models/written_submission.rb [hunk 17]
```diff
@@ -32,6 +33,12 @@ def submit_applications
     end
   end
 
+  def track_applications_submitted
+    user = applications_for_submission.first.user
```
Comment (khanotations) [inline app/models/written_submission.rb]: I think `written_submission` should also have `user` and `sessions` methods

File: app/services/event_service.rb [hunk 18]
```diff
@@ -0,0 +1,19 @@
+class EventService
+  def self.sign_up(user)
+    Event::SignUp.create!(eventable: user)
+  end
+
+  def self.written_application_started(user, sessions)
+    Event::WrittenApplicationStarted.create!(eventable: user,
+      session_names: sessions.map(&:name).join(", "))
```
Comment (khanotations) [inline app/services/event_service.rb]: I think this `sessions.map(&:name).join(", ")` should be standardized somewhere. Let's make a `sessions_to_names` method or something so that we're consistent. We might want `and` or make it a sentence, or something that Intercom can better deal with.

File: config/initializers/analytics_ruby.rb [hunk 19]
```diff
@@ -0,0 +1,5 @@
+Analytics = Segment::Analytics.new({
+    write_key: ENV.fetch('SEGMENT_WRITE_KEY')#,
```
Comment (khanotations) [inline config/initializers/analytics_ruby.rb]: Why `fetch` instead of `ENV[]`?

Comment (JoshBroomberg) [inline config/initializers/analytics_ruby.rb]: http://ruby-doc.org/core-1.9.3/ENV.html#method-c-fetch

Basically, it's a cool helper that either crashes if they key is missing, allows you to supply an inline default or use a block.

File: app/models/user/as_segmentable.rb (hunk 15 above)
Comment (JoshBroomberg) [inline app/models/user/as_segmentable.rb]: Assume that if you resend an identify call with the same id, it does an update. Will confirm but this is obvious behaviour to me.

File: app/models/user/as_segmentable.rb (hunk 14 above)
Comment (JoshBroomberg) [inline app/models/user/as_segmentable.rb]: Literally copy pasted the exact setup from here: https://segment.com/docs/sources/server/ruby/quickstart/

Analytics is an instance.

Seems like they always give the ID as a string, so I interpolated.

File: app/models/user.rb (hunk 16 above)
Comment (JoshBroomberg) [inline app/models/user.rb]: The docs don't specify a specific paradigm for syncing. This catch-all felt like a safer bet than doing a sync at every info change manually?

Comment (JoshBroomberg) [inline app/models/user.rb]: Worth thinking about though

Comment (JoshBroomberg) [inline app/models/user.rb]: Also, unlike in ixstudents, we will not trigger syncs when associated things change. Events will fill that data gap.

**Comment by JoshBroomberg** [comment]
@khanotations all changes made. Even wrote specs for the `EventService`. 

Also linked up the dev segment source to intercom. Works like a charm - both users and events syncing. Also confirmed that If a second call is made to `identify` it updates the user on segment and on intercom.

I believe all we now need to do is:
- Finalise a list of attributes to track
- Decide on when to do updates. (Is after save correct?)
- Decide on which events to track and build them out (very quick)
- Get the front end segment JS working so that it can generate FB pixel and the intercom chat widget for us.

**Comment by JoshBroomberg** [comment]
@khanotations got segment front end integrated. You'll see how I communicate events from back to front is ok but not ideal. I considered using the 'events service as an instance' thing we did in ixstudents but that meant we could only use the service in controllers. 

This method takes advantage of the fact we are persisting events.

**Comment by JoshBroomberg** [comment]
Now depends on #37 because I had to turf intercom gems to make segment intercom work

**Comment by khanotations** [comment]
Note: I pushed to resolve the diff between #37

File: app/controllers/application_controller.rb [hunk 20]
```diff
@@ -17,4 +19,8 @@ def user_not_authorized
     flash[:alert] = "You're in the wrong place, you've been taken back to safety."
     redirect_to(request.referrer || root_path)
   end
+
+  def set_trackable_events
+    @trackable_events = current_user.events.where(tracked: false) if current_user
```
Comment (khanotations) [inline app/controllers/application_controller.rb]: `||=`

File: app/models/event.rb [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+  # (which serve to store event attribute data)
+  # and converts them to json to be stored in 'data' before saving.
+  # This paradigm allows the use of ActiveType attribute fields with
+  # validations on incoming data in event submodels.
+  def assign_json_from_attributes
+    attributes = self.attributes.keys - Event.column_names
+    json = Hash[ attributes.map { |a| [a, self.send(a)] } ]
```
Comment (khanotations) [inline app/models/event.rb]: This mapping could potentially be expensive because it has to recall each method. 

I think this whole method could be one line: 
```
self.data = self.attributes.except(*Event.column_names)
```
Check [except](http://api.rubyonrails.org/classes/Hash.html#method-i-except)

File: app/models/journey.rb [hunk 22]
```diff
@@ -18,11 +19,19 @@ def sanitize_journey_stages
...
+    EventService.written_application_started(user, sessions)
+  end
+
   def user_scope
     User::AsStartedApplicant
   end
 
   def started_user
-    user.as(user_scope)
+    user.try(:as, user_scope)
```
Comment (khanotations) [inline app/models/journey.rb]: When would this ever not have a `user`?

File: app/models/user/as_omniauthable.rb [hunk 23]
```diff
@@ -22,9 +22,16 @@ def self.from_omniauth(auth_data)
...
     user.graduation_year ||= edu_params[1]
     user.save
+
+    set_default_roles(user)
+
     user
   end
 
+  def self.set_default_roles(user)
+    base_user = user.as(User)
```
Comment (khanotations) [inline app/models/user/as_omniauthable.rb]: Why is this necessary?

File: app/views/partials/scripts/segment/_track_events.html.erb [hunk 24]
```diff
@@ -0,0 +1,6 @@
+<script>
+  <% events.each do |event| %>
+    analytics.track("<%= event.event_type %>", <%= event.data.to_json.html_safe %>);
+  <% end %>
+  <% events.update_all(tracked: true) %>
```
Comment (khanotations) [inline app/views/partials/scripts/segment/_track_events.html.erb]: Is running a database query from the view the recommended way to do this..?

File: app/views/layouts/application.html.erb [hunk 25]
```diff
@@ -7,6 +7,7 @@
   <!-- Main css styles -->
   <%= stylesheet_link_tag 'student', media: 'all' %>
   <%= csrf_meta_tags %>
+  <%= render "partials/scripts/segment/analytics" %>
```
Comment (khanotations) [inline app/views/layouts/application.html.erb]: Why is this JS up here instead of right above the other segment stuff?

File: spec/services/event_service_spec.rb [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+  
+  before do
+    event_class.stub(:create!)
+  end
+
+  describe ".sign_up" do
+    let(:event_class) { Event::SignUp }
+    
+    it "creates a sign up event" do
+      expect(event_class).to receive(:create!)
```
Comment (khanotations) [inline spec/services/event_service_spec.rb]: these matchers are so nice

File: spec/services/event_service_spec.rb [hunk 27]
```diff
@@ -0,0 +1,53 @@
...
+
+  context "session based events" do
+    let(:sessions) { build_stubbed_list(:session, 2) }
+    let(:session_names) { sessions.map(&:name).join(", ") }
+    
+    describe ".written_application_started" do
+      let(:event_class) { Event::WrittenApplicationStarted }
+      
+      it "creates a written application started event" do
+        session_names = sessions.map(&:name).join(", ")
```
Comment (khanotations) [inline spec/services/event_service_spec.rb]: did you forget to remove this after using the `let` above?

File: spec/services/event_service_spec.rb [hunk 28]
```diff
@@ -0,0 +1,53 @@
+require 'rails_helper'
```
Comment (khanotations) [inline spec/services/event_service_spec.rb]: these tests are really nice, well done

**Comment by khanotations** [comment]
...my push didn't work :(

but this PR looks great. just small fixes!

File: app/models/interview_submission.rb [hunk 29]
```diff
@@ -23,6 +26,10 @@ def create_interview_response_questions
     end
   end
 
+  def track_interview_submitted
+    EventService.interview_submitted(submitted_user)
+  end
+
   def interview_questions
```
Comment (khanotations) [inline app/models/interview_submission.rb]: now you can do `delegate :interview_questions, to: :submitted_user`

Comment (khanotations) [inline app/models/interview_submission.rb]: unless you feel like this should be a query with some `includes`

**Comment by JoshBroomberg** [comment]
@khanotations I have included some commented code in the `EventService`. The code uses the `hash.new([])`  syntax so is a bit cleaner than the uncommented code. But, when I was doing it I was getting the weirdest error. Looking like some kind of odd pass by reference dependency loop where the hash just wast not working as it should. 

Could you try uncomment the code and run the test suite? If you get the same bug, tests will fail.

**Comment by JoshBroomberg** [comment]
But the conversion to memory based events is now done and tested.

File: app/models/event.rb (hunk 21 above)
Comment (khanotations) [inline app/models/event.rb]: @JoshBroomberg

File: app/services/event_service.rb [hunk 30]
```diff
@@ -0,0 +1,76 @@
...
+  def self.interview_submitted(user)
+    Event::InterviewSubmitted.create!(user: user)
+  end
+
+  # This method is used in an after create on the event model.
+  # This ensures no model can be created without being stored using
+  # this method.
+  def self.temp_store(event)
+    return unless event.persisted?
+    events[event.user_id] = [] unless events[event.user_id]
```
Comment (khanotations) [inline app/services/event_service.rb]: `events[event.user_id] ||= []`

Even better is `(events[event.user_id] ||= []) << event`

File: app/services/event_service.rb [hunk 31]
```diff
@@ -0,0 +1,76 @@
...
+  end
+
+  private
+
+  def self.events
+    @@events_hash ||= {}
+  end
+
+  # NO MERGE
+  # Alternative code:
```
Comment (khanotations) [inline app/services/event_service.rb]: I like this better

File: app/views/partials/scripts/segment/_track_events.html.erb [hunk 32]
```diff
@@ -0,0 +1,7 @@
+<script>
+  <% events = EventService.temp_stored_events_for(current_user) %>
```
Comment (khanotations) [inline app/views/partials/scripts/segment/_track_events.html.erb]: I would just get rid of this variable: `EventService.temp_stored_events_for(current_user).each do`

File: app/services/event_service.rb [hunk 33]
```diff
@@ -0,0 +1,76 @@
...
+
+  def self.temp_stored_events_for(user)
+    events[user.id] || []
+  end
+
+  def self.clear_temp_stored_events_for(user)
+    events[user.id] = []
+  end
+
+  def self.clear_all_temp_stored_events
```
Comment (khanotations) [inline app/services/event_service.rb]: If this method is only used in tests, it should go in `spec/support` or in the factory as a convenience method. Then you can also avoid testing it.

File: app/services/event_service.rb [hunk 34]
```diff
@@ -0,0 +1,76 @@
...
+  end
+
+  def self.interview_submitted(user)
+    Event::InterviewSubmitted.create!(user: user)
+  end
+
+  # This method is used in an after create on the event model.
+  # This ensures no model can be created without being stored using
+  # this method.
+  def self.temp_store(event)
```
Comment (khanotations) [inline app/services/event_service.rb]: I'm not a fan of the `temp_` naming. I'd think `store_untracked`, `untracked_events_for`, `clear_untracked_events_for`. Even `untracked_for` and `clear_untracked_for` since it's already on `EventService`.

File: app/models/event.rb [hunk 35]
```diff
@@ -0,0 +1,32 @@
...
+
+  enum event_type: [
+    Constants::Events::SIGN_UP,
+    Constants::Events::WRITTEN_APPLICATION_STARTED,
+    Constants::Events::WRITTEN_APPLICATION_SUBMITTED,
+    Constants::Events::INTERVIEW_SUBMITTED
+  ]
+  
+  before_save :assign_json_from_attributes
+  after_create :store_event
```
Comment (khanotations) [inline app/models/event.rb]: `store_untracked_event`

File: app/services/event_service.rb (hunk 34 above)
Comment (khanotations) [inline app/services/event_service.rb]: Also these methods feel wrong in this class. Is there a better place to put them? If it's all in one request cycle, should it go in a helper?

File: spec/services/event_service_spec.rb [hunk 36]
```diff
@@ -0,0 +1,117 @@
...
+        it "clears specified user's events" do
+          expect(EventService.temp_stored_events_for(user)).to match_array([])
+        end
+
+        it "maintains other users' events" do
+          expect(EventService.temp_stored_events_for(second_user)).to match_array([second_event])
+        end
+      end
+
+      describe ".clear_all_temp_stored_events" do
```
Comment (khanotations) [inline spec/services/event_service_spec.rb]: If this method is only used in tests, it shouldn't go on the model, and doesn't need to be tested.

File: spec/services/event_service_spec.rb [hunk 37]
```diff
@@ -0,0 +1,117 @@
...
+      describe ".temp_stored_events_for" do
+        it "returns user events" do
+          expect(EventService.temp_stored_events_for(user)).to match_array([event])
+        end
+      end
+
+      describe ".clear_temp_stored_events_for" do
+        before { EventService.clear_temp_stored_events_for(user) }
+        it "clears specified user's events" do
+          expect(EventService.temp_stored_events_for(user)).to match_array([])
```
Comment (khanotations) [inline spec/services/event_service_spec.rb]: `be_empty`?

File: spec/services/event_service_spec.rb [hunk 38]
```diff
@@ -0,0 +1,117 @@
...
+        it "stores event" do
+          expect(EventService.temp_stored_events_for(user)).to match_array([])
+        end
+      end
+    end
+    
+    context "no events stored" do
+      describe ".temp_stored_events_for" do
+        it "returns empty array" do
+          expect(EventService.temp_stored_events_for(user)).to match_array([])
```
Comment (khanotations) [inline spec/services/event_service_spec.rb]: `be_empty`

File: app/services/event_service.rb (hunk 31 above)
Comment (JoshBroomberg) [inline app/services/event_service.rb]: Something extremely odd is happening with that code. Behaving in a way I can;t explain.

Comment (JoshBroomberg) [inline app/services/event_service.rb]: So you can;t have it

File: app/services/event_service.rb (hunk 33 above)
Comment (JoshBroomberg) [inline app/services/event_service.rb]: In order to access the hash from outside the class, I need to do weird stuff to the variable. Not safe programming. Gonna leave it like this.

File: app/services/event_service.rb (hunk 34 above)
Comment (JoshBroomberg) [inline app/services/event_service.rb]: Did the renaming but I am not sure what you mean re the file. This feels like the exact place for these methods?

File: spec/services/event_service_spec.rb (hunk 36 above)
Comment (JoshBroomberg) [inline spec/services/event_service_spec.rb]: See up

File: app/models/event.rb (hunk 21 above)
Comment (JoshBroomberg) [inline app/models/event.rb]: wow, nice

File: app/models/interview_submission.rb [hunk 39]
```diff
@@ -4,6 +4,9 @@ class InterviewSubmission < ActiveType::Object
   
   before_save :validate_interview_response
   after_save :create_interview_response_questions
+  after_save :track_interview_submitted
```
Comment (khanotations) [inline app/models/interview_submission.rb]: are these after saves or after creates?

File: app/models/journey.rb [hunk 40]
```diff
@@ -3,6 +3,7 @@ class Journey < ActiveType::Object
   nests_many :journey_stages, scope: Intent::JourneyStage, default: [],
     reject_if: -> (attributes) { attributes[:session_id].blank? } 
   before_save :sanitize_journey_stages
+  after_save :track_applications_started
```
Comment (khanotations) [inline app/models/journey.rb]: also here should this be after create?

File: app/services/event_service.rb [hunk 41]
```diff
@@ -0,0 +1,50 @@
...
+    @@events_hash = {}
+  end
+
+  private
+
+  def self.events
+    @@events_hash ||= {}
+  end
+
+  def self.sessions_to_string(sessions)
```
Comment (khanotations) [inline app/services/event_service.rb]: do you think this should be a method on the Session class so we do it consistently always? or do you think it's good to make it unique to evebts?

File: app/services/event_service.rb [hunk 42]
```diff
@@ -0,0 +1,50 @@
...
+
+  def self.untracked_events_for(user)
+    events[user.id] || []
+  end
+
+  def self.clear_untracked_events_for(user)
+    events[user.id] = []
+  end
+
+  def self.clear_all_untracked_events
```
Comment (khanotations) [inline app/services/event_service.rb]: we were going to move this to a testing helper right?

File: Gemfile [hunk 43]
```diff
@@ -16,6 +16,7 @@ gem "pry-rails", "~> 0.3.4" # Improved production console.
 # External services
 gem "aws-sdk", "< 2.0" # S3 storage.
 gem "omniauth-facebook", "~> 3.0.0" # Facebook login.
+gem "analytics-ruby", "~> 2.0.0", require: "segment/analytics"
```
Comment (khanotations) [inline Gemfile]: what is this

File: app/models/interview_submission.rb (hunk 39 above)
Comment (JoshBroomberg) [inline app/models/interview_submission.rb]: Scroll slightly up from [here](https://github.com/makandra/active_type#activetyperecord) to see the AT docs on this. I don't think AT objects have create hooks because those are DB levels.

File: app/models/journey.rb (hunk 40 above)
Comment (JoshBroomberg) [inline app/models/journey.rb]: See above

File: app/services/event_service.rb (hunk 41 above)
Comment (JoshBroomberg) [inline app/services/event_service.rb]: I like me consistency

File: app/services/event_service.rb (hunk 42 above)
Comment (JoshBroomberg) [inline app/services/event_service.rb]: No, I couldn't move it because I would have had to screw with the privacy of the array to access it from outside.

File: Gemfile (hunk 43 above)
Comment (JoshBroomberg) [inline Gemfile]: This is segment

File: app/models/user/as_segmentable.rb (hunk 15 above)
Comment (JoshBroomberg) [inline app/models/user/as_segmentable.rb]: Can confirm that this updates the user with the new value

**Comment by JoshBroomberg** [comment]
ok @khanotations, this is good for review again.
- Did some general clean up, addressing some of yours comments
- Tested the segment integration and the Facebook/intercom linkages. All good on this front. 
- Cleaned up the roles system

**Comment by JoshBroomberg** [comment]
The todo list in the description still stands. Let's talk about them asap.

File: app/services/event_service.rb [hunk 44]
```diff
@@ -0,0 +1,46 @@
...
+    @@events_hash = {}
+  end
+
+  private
+
+  def self.events
+    @@events_hash ||= {}
```
Comment (khanotations) [inline app/services/event_service.rb]: Why aren't we doing `Hash.new([])`?
